### PR TITLE
feat: profiling v2 [MD-27]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Profiling V2 ([Project Doc](https://github.com/azhou-determined/notes/blob/master/profiling-v2/profiling-v2.md))

Individual commits should have already been reviewed, this is the final feature branch to main PR.

Major changes:

- Functionality for timing metrics in the Determined profiler was dropped. Determined profiler is now only responsible for system metrics.
- Determined Profiler now lives in Core API. 
- Determined profiler's system metrics now use backend's generic metrics framework

## Test Plan

This feature should be tested across a few different Trial APIs

### PyTorch

For `mnist_pytorch`:
- Change `determined/examples/tutorials/mnist_pytorch/train.py` to add `profiling_enabled`:
```
trainer.fit(max_length=max_length, latest_checkpoint=latest_checkpoint, profiling_enabled=True)
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.

### TF Keras
- In `determined/examples/computer_vision/iris_tf_keras/distributed.yaml`, enabled profiling:
```
profiling:
  enabled: true
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.
image

### Core API
- Submit a Core API experiment:

```python
import time
import logging

import determined as det
from determined import core


def main(core_context):
    core_context.profiler.on(sampling_interval=0.1, samples_per_report=10)
    for batch in range(30):
        steps_completed = batch + 1
        if steps_completed % 5 == 0:
            core_context.train.report_training_metrics(
                steps_completed=steps_completed, metrics={"x": batch}
            )
        if steps_completed % 10 == 0:
            core_context.train.report_validation_metrics(steps_completed=steps_completed, metrics={"x": batch})
        time.sleep(1)
    core_context.profiler.off()


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format=det.LOG_FORMAT)
    with core.init() as core_context:
        main(core_context=core_context)
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by gt2345 — APPROVED** [review]
WebUI stamp

File: docs/model-dev-guide/profiling.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+inherent to the task, unnecessary resource consumption can be caused by suboptimal code or bugs.
+Thus, achieving optimal model performance requires an understanding of how your model interacts with
+the system's computational resources.
+
+Profiling collects metrics on how computational resources like CPU, GPU, and memory are being
+utilized during a training job. It can reveal patterns in resource utilization that indicate
+performance bottlenecks and pinpoint areas of the code or pipeline that are causing slowdowns or
+inefficiencies.
+
+Profiling a training job can be instrumented at many different layers, from generic system-level
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:19]:
```suggestion
A training job can be instrumented at many different layers, from generic system-level
```
or
```suggestion
A training job can be profiled at many different layers, from generic system-level
```

File: docs/model-dev-guide/profiling.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+The Determined profiler collects a set of system metrics throughout an experiment which can be
+visualized in the WebUI under the experiment's "Profiler" tab. It is supported for all training
+APIs, but is not enabled by default.
+
+Visit :ref:`core-profiler` to find out how to enable and configure the Determined profiler for your
+experiment.
+
+The following system metrics are tracked:
+
+-  *GPU utilization (percent)*: current utilization of a GPU device
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:56]: I'm vaguely uncomfortable with "current", and I can't pinpoint the concept I think it's meant to indicate. I believe that the profiler aggregates average utilization over each interval. But I think that not specifying the exact method and letting users read some docstrings or code if they really care is a fine choice.

So I guess my recommendation is to just drop the word "current" wherever it appears.

Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:56]: I like the inclusion of units, BTW.

File: docs/model-dev-guide/profiling.rst [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+-  *GPU free memory (bytes)*: current amount of free memory available on a GPU device
+-  *Network throughput - sent (bytes/s)*: change rate of number of bytes sent system-wide
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:58]: I don't know what "change rate" means. Here, units are good enough that I'd probably understand if it just said "bytes sent syntemwide". The [wikipedia page](https://en.wikipedia.org/wiki/Data-rate_units) about units doesn't give me any words I like.

File: docs/model-dev-guide/profiling.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+   system
+
+For distributed training, these metrics are collected for every agent. The data is broken down by
+agent, and GPU metrics can be further broken down by GPU.
+
+.. note::
+
+   System Metrics record agent-level metrics, so when there are multiple experiments on the same
+   agent, it is difficult to analyze. It is recommended that profiling is done with only a single
+   experiment per agent.
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:75]: This could be a place to reference how this can be configured.

File: docs/model-dev-guide/profiling.rst [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+   training APIs that provides fine-grained metrics specific to your model.
+-  :ref:`Prometheus/Grafana integration <how-to-profiling-prom-grafana>` can be set up to track
+   detailed hardware metrics and monitor overall cluster health.
+
+.. _how-to-profiling:
+
+.. _how-to-profiling-det-profiler:
+
+*********************
+ Determined Profiler
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:35]: Is it appropriate to provide references here for how profiler metrics can be inspected?

File: docs/model-dev-guide/profiling.rst [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+.. note::
+
+   System Metrics record agent-level metrics, so when there are multiple experiments on the same
+   agent, it is difficult to analyze. It is recommended that profiling is done with only a single
+   experiment per agent.
+
+.. _how-to-profiling-native-profilers:
+
+***************************
+ Native Training Profilers
```
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:80]: This is nice. I think that overall, this doc is set up well.

File: docs/reference/experiment-config-reference.rst [hunk 7]
```diff
@@ -1483,38 +1483,19 @@ explicitly specified, the master will automatically generate an experiment seed.
...
+The ``profiling`` section specifies configuration options for the Determined system metrics
+profiler. See :ref:`how-to-profiling` for a more detailed walkthrough.
 
 ``profiling``
 =============
 
-Optional. Profiling is supported for all frameworks, though timings are only collected for
-``PyTorchTrial``. Profiles are collected for a maximum of 5 minutes, regardless of the settings
-below.
+Optional. Defaults to false.
```
Comment (wes-turner) [inline docs/reference/experiment-config-reference.rst:1492]: What does it do?

File: docs/.redirects/redirects.json [hunk 8]
```diff
@@ -1,6 +1,7 @@
 {
     "reference/python-sdk": "python-sdk/python-sdk.html",
     "reference/training/experiment-config-reference": "../experiment-config-reference.html",
+    "model-dev-guide/dtrain/optimize-training": "../profiling.html",
```
Comment (wes-turner) [inline docs/.redirects/redirects.json:4]: Oh -- nice thinking. I should remember this.

File: harness/determined/core/_profiler.py [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+    def _close(self) -> None:
+        pass
+
+
+def _average_metric_samples_depth_one(metric_samples: List[Dict[str, Any]]) -> Dict[str, Any]:
+    """Helper method to merge a list of dictionary averaging their values by their keys.
+
+    Supports up to 1 level of nesting. Returns a single merged dictionary where the values are
+    averaged across all dictionaries in the given list by key.
+    # TODO (anda): find a cleaner way to do this.
```
Comment (wes-turner) [inline harness/determined/core/_profiler.py:166]: If this is important to you, could you please create a ticket and reference the ticket in the TODO instead of yourself? And if it's not important enough for that, then please remove the TODO.

File: harness/determined/core/_profiler.py [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+    """
+    aggregated_metrics: Dict[str, Any] = {}
+    for sample in metric_samples:
+        for k, v in sample.items():
+            if isinstance(v, dict):
+                aggregated_metrics[k] = aggregated_metrics.get(k, {})
```
Comment (wes-turner) [inline harness/determined/core/_profiler.py:172]: nit: little clearer with a `defaultdict`, I think.

File: harness/determined/core/_profiler.py [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+                    "gpu_util": 0.23,
+                    "gpu_free_memory": 234.56,
+                }
+            }
+        """
+        pass
+
+
+class _Network(_MetricGroupCollector):
+    group = "network"
```
Comment (wes-turner) [inline harness/determined/core/_profiler.py:246]: I'm a little surprised this works. I'd have expected that it had to look like
```python
@property
def group:
  return "network"
```

My typechecker doesn't like it, either.

File: harness/determined/core/_profiler.py [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+        """Initialize the pynvml library and validate methods.
+
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+
+        If any NVML method fails for any GPU device, no GPU metrics will be collected for
+        all GPU devices.
+        """
+        try:
+            assert pynvml
```
Comment (wes-turner) [inline harness/determined/core/_profiler.py:368]: nit: could you please pull this outside the try? Makes my IDE happier for the except block.

File: harness/tests/core/test_profiler.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    expected_metrics = {
+        "cpu_util_simple": mock_cpu_util,
+    }
+
+    cpu_collector.sample_metrics()
+    assert len(cpu_collector.metric_samples) == 1
+    assert cpu_collector.metric_samples[0] == expected_metrics
+
+
+@pytest.mark.skipif(profiler.pynvml is None, reason="pynvml required for test")
```
Comment (wes-turner) [inline harness/tests/core/test_profiler.py:148]: How do we keep this tested in CI?

File: harness/tests/core/test_profiler.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+            "gpu_util": mock_gpu_util[i],
+        }
+
+    gpu_collector.sample_metrics()
+    assert len(gpu_collector.metric_samples) == 1
+    assert gpu_collector.metric_samples[0] == expected_metrics
+
+
+def test_average_metric_samples() -> None:
+    test_metrics_flat = [
```
Comment (wes-turner) [inline harness/tests/core/test_profiler.py:216]: Each test, one behavior.

File: docs/model-dev-guide/profiling.rst (hunk 4 above)
Comment (azhou-determined) [reply docs/model-dev-guide/profiling.rst:75]: this can't really be configured. i left this warning from the previous doc because it's still relevant, but it has to do with being aware that there could be other experiments using the same agent as your experiment.

File: harness/tests/core/test_profiler.py (hunk 13 above)
Comment (azhou-determined) [reply harness/tests/core/test_profiler.py:148]: this is tested by `test-unit-harness-gpu`

File: harness/tests/core/test_profiler.py (hunk 14 above)
Comment (azhou-determined) [reply harness/tests/core/test_profiler.py:216]: sure done

File: docs/reference/experiment-config-reference.rst (hunk 7 above)
Comment (wes-turner) [reply docs/reference/experiment-config-reference.rst]: Not being purposefully dense, I can't differentiate between:
```yaml
profiling:
  profiler: <val>
  enabled: <val>
```

File: docs/model-dev-guide/profiling.rst (hunk 4 above)
Comment (wes-turner) [reply docs/model-dev-guide/profiling.rst:73]: I think it would be frustrating to me to read this note and then be unable to figure out how to do the thing it recommends. Let's chat about it.

File: docs/model-dev-guide/profiling.rst (hunk 5 above)
Comment (wes-turner) [reply docs/model-dev-guide/profiling.rst:35]: Mmm. I missed this part.

> The Determined profiler collects a set of system metrics throughout an experiment which can be visualized in the WebUI under the experiment’s “Profiler” tab. It is supported for all training APIs, but is not enabled by default.

My fault. Like usual, I'd like someone else to blame when a thing was my fault. In that respect, could / should it have been clearer?

File: harness/determined/core/_profiler.py (hunk 11 above)
Comment (wes-turner) [reply harness/determined/core/_profiler.py]: I'm a little unsettled by this. I don't think the implemented code works, either, and that it hasn't resulted in failed tests makes me wonder if something is wrong with tests.

currently implemented:
```python
    def group(self) -> str:
        return "network"
```

File: harness/determined/core/_profiler.py (hunk 10 above)
Comment (wes-turner) [reply harness/determined/core/_profiler.py:172]:
```python
aggregated_metrics = defaultdict(int)
aggregated_metrics = defaultdict(defaultdict(int))
```

File: e2e_tests/tests/experiment/test_profiling.py [hunk 15]
```diff
@@ -15,12 +15,10 @@
 @pytest.mark.e2e_slurm_gpu
 @pytest.mark.timeout(30 * 60)
 @pytest.mark.parametrize(
-    "model_def,timings_enabled",
-    [
-        (conf.fixtures_path("mnist_pytorch"), True),
-    ],
+    "model_def",
```
Comment (wes-turner) [inline e2e_tests/tests/experiment/test_profiling.py:18]: nit: I don't understand the parameterization. I think it's cleaner without it

File: docs/model-dev-guide/profiling.rst (hunk 4 above)
Comment (azhou-determined) [reply docs/model-dev-guide/profiling.rst]: removed this recommendation

File: docs/reference/experiment-config-reference.rst (hunk 7 above)
Comment (azhou-determined) [reply docs/reference/experiment-config-reference.rst]: removed this section

File: harness/determined/core/_profiler.py (hunk 9 above)
Comment (azhou-determined) [reply harness/determined/core/_profiler.py:166]: i've created the ticket https://hpe-aiatscale.atlassian.net/browse/MD-338. i will add it to the comment but don't want to commit right now because i'm waiting for a longrunning CI to finish.

File: harness/determined/core/_profiler.py (hunk 11 above)
Comment (azhou-determined) [reply harness/determined/core/_profiler.py]: fixed

File: harness/determined/core/_profiler.py (hunk 10 above)
Comment (azhou-determined) [reply harness/determined/core/_profiler.py:172]: eh, don't love the way this reads. `.get(k, {})` is easily readable IMO. i do think `defaultdict` is best-practices way of doing this, but since i have a ticket to refactor this method anyway, i don't think it's worth it.

File: e2e_tests/tests/experiment/test_profiling.py (hunk 15 above)
Comment (azhou-determined) [reply e2e_tests/tests/experiment/test_profiling.py:18]: done

**Review by wes-turner — APPROVED** [review]
LGTM

File: master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql
Comment (jgongd) [inline master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql:1]: The most recent migration file on 0.29.1 was named [20240306164812_remove-rm-name-expconf.tx.up.sql](https://github.com/determined-ai/determined/blob/0.29.1/master/static/migrations/20240306164812_remove-rm-name-expconf.tx.up.sql).

I suggest moving these files to the top to avoid any migration issues when users update. 
`/migration-move-to-top.sh my-migration-name`

File: proto/src/determined/api/v1/experiment.proto [hunk 16]
```diff
@@ -631,6 +631,8 @@ enum MetricType {
   METRIC_TYPE_TRAINING = 1;
   // For metrics emitted during validation.
   METRIC_TYPE_VALIDATION = 2;
+  // For metrics emitted during profiling.
+  METRIC_TYPE_PROFILING = 3;
```
Comment (jgongd) [inline proto/src/determined/api/v1/experiment.proto:635]: https://github.com/determined-ai/determined/blob/main/master/pkg/model/metrics.go#L41
https://github.com/determined-ai/determined/blob/main/master/pkg/protoutils/protoconverter/converter.go#L169

These functions may need to be updated to work with the new type.

Comment (azhou-determined) [reply proto/src/determined/api/v1/experiment.proto:635]: the READ part of getting profiling metrics from generic metrics isn't implemented in this feature, only the write. so i think it's appropriate for the time being that those functions return "unspecified" or "unknown", because at this point in time, profiling metrics should not be known to those APIs.

File: master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql
Comment (azhou-determined) [reply master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql:1]: done!

**Review by jgongd — APPROVED** [review]
